File: edit_distance/security_ed.py
import math

import matplotlib.pyplot as plt
import numpy as np
from scipy.special import comb

M = 128
lowers = np.zeros(M + 1)
for m in range(len(lowers)):
    lowers[m] = pow(2, m)

Mp = 128
points128 = np.zeros(M + 1)
flag = True
for m in range(M + 1):
    tmp = 0
    for Mr in range(0, m + 1):
        tmp += comb(m, Mr) * comb(Mp, Mr)
    points128[m] = tmp
    if flag and points128[m] < lowers[m]:
        print("128", m)
        flag = False

Mp = 96
points96 = np.zeros(M + 1)
flag = True
for m in range(M + 1):
    tmp = 0
    for Mr in range(0, m + 1):
        tmp += comb(m, Mr) * comb(Mp, Mr)
    points96[m] = tmp
    if flag and points96[m] < lowers[m]:
        print("96", m)
        flag = False

Mp = 64
points64 = np.zeros(M + 1)
flag = True
for m in range(M + 1):
    tmp = 0
    for Mr in range(0, m + 1):
        tmp += comb(m, Mr) * comb(Mp, Mr)
    points64[m] = tmp
    if flag and points64[m] < lowers[m]:
        print("64", m)
        flag = False

Mp = 48
points48 = np.zeros(M + 1)
flag = True
for m in range(M + 1):
    tmp = 0
    for Mr in range(0, m + 1):
        tmp += comb(m, Mr) * comb(Mp, Mr)
    points48[m] = tmp
    if flag and points48[m] < lowers[m]:
        print("48", m)
        flag = False

Mp = 32
points32 = np.zeros(M + 1)
flag = True
for m in range(M + 1):
    tmp = 0
    for Mr in range(0, m + 1):
        tmp += comb(m, Mr) * comb(Mp, Mr)
    points32[m] = tmp
    if flag and points32[m] < lowers[m]:
        print("32", m)
        flag = False


plt.close()
plt.figure()
plt.grid(linestyle="--")
plt.xlabel("$M$", fontsize=12, fontweight='bold')
plt.ylabel("The number of guesses", fontsize=12, fontweight='bold')
plt.yscale('log')
# plt.scatter(range(len(points224)), points224, s=5, marker=".", color="b", linewidth=2, label="$N_{guess}$, $N_p=224$")
# plt.scatter(range(len(points192)), points192, s=5, marker=">", color="g", linewidth=2, label="$N_{guess}$, $N_p=192$")
# plt.scatter(range(len(points160)), points160, s=5, marker="<", color="k", linewidth=2, label="$N_{guess}$, $N_p=160$")
# plt.scatter(range(len(points128)), points128, s=5, marker="v", color="c", linewidth=2, label="$N_{guess}$, $N_p=128$")
# plt.scatter(range(len(points64)), points64, s=5, marker="^", color="m", linewidth=2, label="$N_{guess}$, $N_p=64$")
# plt.scatter(range(len(lowers)), lowers, s=5, marker="+", color="r", linewidth=2, label="$2^{2N_r}$")

x0 = 104
y0 = points32[x0]
plt.scatter(x0, y0, color="red")
# plt.plot([x0, x0], [y0, 0], "r--")
plt.annotate("$(%d,%.2e)$" % (x0, y0), xy=(x0, y0), xytext=(-60, -40), textcoords='offset points', fontsize=10,
             arrowprops=dict(arrowstyle='->', connectionstyle='arc3,rad=.2'))
plt.plot(range(len(points128)), points128, linestyle='--', color="b", linewidth=2, label="$M_{guess}^{SA}$, $M_p=128$")
plt.plot(range(len(points96)), points96, linestyle='-', color="m", linewidth=2, label="$M_{guess}^{SA}$, $M_p=96$")
plt.plot(range(len(points64)), points64, linestyle=':', color="c", linewidth=2, label="$M_{guess}^{SA}$, $M_p=64$")
plt.plot(range(len(points48)), points48, linestyle='-.', color="g", linewidth=2, label="$M_{guess}^{SA}$, $M_p=48$")
plt.plot(range(len(points32)), points32, linestyle='--', color="k", linewidth=2, label="$M_{guess}^{SA}$, $M_p=32$")
plt.plot(range(len(lowers)), lowers, color="r", linewidth=2, label="$2^M$")
plt.ylim(1.66223336e-11, 1.268869e+97)

plt.legend(loc='upper left')
leg_text = plt.gca().get_legend().get_texts()
plt.setp(leg_text, fontsize=10, fontweight='bold')
plt.savefig('./evaluations/nAllPlot_ed.pdf', format='pdf')
plt.show()

File: sorting_index/plot/security_ed.m
clear; close all; clc
textSize = 14;
linewid = 1;

points128 = [];
points96 = [];
points64 = [];
points48 = [];
points32 = [];
lowers = [];

for i=1:129
    lowers(i) = 2^(i-1);
end

mp = 128;
for i=1:129
    tmp = 0;
    for mr=1:i
        if (mr-1<=mp) && (mr-1<=i-1)
            tmp = tmp + nchoosek(i-1,mr-1) * nchoosek(mp,mr-1);
        end
    end
    points128(i) = tmp;
end

mp = 96;
for i=1:129
    tmp = 0;
    for mr=1:i
        if (mr-1<=mp) && (mr-1<=i-1)
            tmp = tmp + nchoosek(i-1,mr-1) * nchoosek(mp,mr-1);
        end
    end
    points96(i) = tmp;
end

mp = 64;
for i=1:129
    tmp = 0;
    for mr=1:i
        if (mr-1<=mp) && (mr-1<=i-1)
            tmp = tmp + nchoosek(i-1,mr-1) * nchoosek(mp,mr-1);
        end
    end
    points64(i) = tmp;
end

mp = 48;
for i=1:129
    tmp = 0;
    for mr=1:i
        if (mr-1<=mp) && (mr-1<=i-1)
            tmp = tmp + nchoosek(i-1,mr-1) * nchoosek(mp,mr-1);
        end
    end
    points48(i) = tmp;
end

mp = 32;
for i=1:129
    tmp = 0;
    for mr=1:i
        if (mr-1<=mp) && (mr-1<=i-1)
            tmp = tmp + nchoosek(i-1,mr-1) * nchoosek(mp,mr-1);
        end
    end
    points32(i) = tmp;
end

figure1 = figure;
plot([1:129], points128, 'b--', 'LineWidth',2)
hold on
plot([1:129], points96, 'm-', 'LineWidth',2)
hold on
plot([1:129], points64, 'c:', 'LineWidth',2)
hold on
plot([1:129], points48, 'g-.', 'LineWidth',2)
hold on
plot([1:129], points32, 'k--', 'LineWidth',2)
hold on
plot([1:129], lowers, 'r-', 'LineWidth',2)
hold on
plot(104,1.3502058579275469e+31,'ro','LineWidth',2);
hold on;

ax = gca;
ax.YAxis.FontSize = 13;
ax.XAxis.FontSize = 13;
ax.YAxis.FontWeight = 'bold';
ax.XAxis.FontWeight = 'bold';
ax.YScale='log';
ax.YTick=[0, 1e+20, 1e+40, 1e+60, 1e+80, 1e+100, 1e+120];

ylim([0, 1e+120]);
legend('M_{guess}^{SA}, M_p=128','M_{guess}^{SA}, M_p=96',...
    'M_{guess}^{SA}, M_p=64', 'M_{guess}^{SA}, M_p=48',...
    'M_{guess}^{SA}, M_p=32', '2^M', ...
    'Position',[0.151190476190476 0.630952380952381 0.653571428571428 0.278571428571429],...
    'NumColumns',2, 'FontSize',textSize,...
    'FontWeight','bold', 'Location','northwest')
xlabel('M', 'FontSize', 15, 'FontWeight','bold')
ylabel('The number of guesses', 'FontSize',15, 'FontWeight','bold')
grid on

% 创建 arrow
annotation(figure1,'arrow',[0.653571428571429 0.704761904761904],...
    [0.233333333333333 0.325396825396826],'LineWidth',1);

% 创建 textbox
annotation(figure1,'textbox',...
    [0.516666666666666 0.142857142857143 0.273809523809524 0.0968253968253976],...
    'String',{'(104, 1.35e+31)'},...
    'LineStyle','none',...
    'FontWeight','bold',...
    'FontSize',14,...
    'FitBoxToText','off');

print('nAllPlot_ed','-depsc')